File: swift/Sources/AmenClient/Webhooks/Handler.swift
import Foundation

public struct WebhookEvent: Sendable { public let id: String; public let type: String?; public let timestamp: String?; public let data: [String: Any]; public let raw: Data
    public init(id: String, type: String?, timestamp: String?, data: [String: Any], raw: Data) { self.id = id; self.type = type; self.timestamp = timestamp; self.data = data; self.raw = raw } }
public struct WebhookResult: Sendable { public let status: Int; public let body: [String: Any] }

/// Framework-agnostic (Vapor, Hummingbird, …): call `handle(headers:rawBody:)` with the RAW body bytes.
/// Verifies over `timestamp + "." + rawBody`, de-duplicates by the top-level event `timestamp`, then dispatches.
/// Keep `onEvent` fast; queue heavy work.
public final class WebhookHandler: @unchecked Sendable {
    let secret: String; let onEvent: (WebhookEvent) async -> Void
    var seen = Set<String>(); let lock = NSLock()
    public init(secret: String, onEvent: @escaping (WebhookEvent) async -> Void) { self.secret = secret; self.onEvent = onEvent }

    public func handle(headers: [String: String], rawBody: Data) async -> WebhookResult {
        func header(_ name: String) -> String? { headers.first { $0.key.caseInsensitiveCompare(name) == .orderedSame }?.value }
        let sig = header(WebhookSignature.signatureHeader)
        let timestamp = header(WebhookSignature.timestampHeader)
        guard WebhookSignature.verify(secret: secret, timestamp: timestamp, rawBody: rawBody, received: sig) else { return WebhookResult(status: 401, body: ["error": "invalid signature"]) }
        guard let data = (try? JSONSerialization.jsonObject(with: rawBody)) as? [String: Any] else { return WebhookResult(status: 400, body: ["error": "invalid json"]) }
        let type = (data["event"] as? String) ?? header(WebhookSignature.eventHeader)
        // No event id in the body — de-dupe on the top-level event `timestamp` (fall back to the header).
        let id = (data["timestamp"] as? String) ?? timestamp ?? ""
        if !id.isEmpty {
            lock.lock(); let dup = seen.contains(id); if !dup { seen.insert(id) }; lock.unlock()
            if dup { return WebhookResult(status: 200, body: ["ok": true, "duplicate": true]) }
        }
        await onEvent(WebhookEvent(id: id, type: type, timestamp: (data["timestamp"] as? String) ?? timestamp, data: data, raw: rawBody))
        return WebhookResult(status: 200, body: ["ok": true])
    }
}

File: flutter/amen_client/lib/src/webhooks/handler.dart
import 'dart:convert';
import 'verify.dart';

class WebhookEvent {
  final String id;
  final String? type;
  final Map<String, dynamic> data;
  final List<int> raw;
  WebhookEvent(this.id, this.type, this.data, this.raw);
}

class WebhookResult {
  final int status;
  final Map<String, Object?> body;
  WebhookResult(this.status, this.body);
}

/// Framework-agnostic (shelf, dart_frog, …): call [handle] with the headers and
/// RAW body bytes. Verifies first, de-duplicates by event timestamp, then
/// dispatches. Keep [onEvent] fast; queue heavy work.
class WebhookHandler {
  final String secret;
  final Future<void> Function(WebhookEvent) onEvent;
  final Set<String> seen;
  WebhookHandler(this.secret, this.onEvent, {Set<String>? seen})
      : seen = seen ?? {};

  String? _header(Map<String, String> headers, String name) => headers.entries
      .where((e) => e.key.toLowerCase() == name)
      .map((e) => e.value)
      .firstOrNull;

  Future<WebhookResult> handle(
    Map<String, String> headers,
    List<int> rawBody,
  ) async {
    final sig = _header(headers, signatureHeader);
    final timestamp = _header(headers, timestampHeader);
    if (!verifySignature(secret, timestamp, rawBody, sig)) {
      return WebhookResult(401, {'error': 'invalid signature'});
    }
    Map<String, dynamic> data;
    try {
      data = (jsonDecode(utf8.decode(rawBody)) as Map).cast();
    } catch (_) {
      return WebhookResult(400, {'error': 'invalid json'});
    }
    // No event id in the body — de-dupe on the event timestamp, falling back to
    // the header timestamp.
    final id = (data['timestamp'] ?? timestamp ?? '').toString();
    if (id.isNotEmpty && seen.contains(id)) {
      return WebhookResult(200, {'ok': true, 'duplicate': true});
    }
    if (id.isNotEmpty) seen.add(id);
    final type = (data['event'] ?? _header(headers, eventHeader)) as String?;
    await onEvent(WebhookEvent(id, type, data, rawBody));
    return WebhookResult(200, {'ok': true});
  }
}
